Sage/carbon
# fix(flat-table): ensure th elements within sticky thead overlap tbody elements when scrolling FE-5095

fix #5459
fix #5576

### Proposed behaviour


Increments the `z-index` of `th` elements within `thead` when `hasStickyHead` prop is set. This is
because when `tr` elements are interactive or have `subRows` they set the `z-index` of the child
`td` elements to the same `z-index` and are defined later in the DOM so overlap.

Adds additional support for rendering both left and right hand side sticky columns using the
`FlatTableRowHeader` component. A new `stickyAlignment` prop has been added which defaults to
`left`. When `right` is passed it will lock the given column and any columns following it whilst the
remaining columns can be scrolled.

![image](https://user-images.githubusercontent.com/44157880/197814817-c9502518-87fa-49c5-84a5-be3880c268ca.png)

![image](https://user-images.githubusercontent.com/44157880/197814874-f8acf163-6f8c-46a7-8806-2ba5cc926b48.png)

![expandable sticky cols](https://user-images.githubusercontent.com/44157880/197815942-37241fe7-5d77-4f15-9eab-10b849d085c3.gif)

![sticky cols](https://user-images.githubusercontent.com/44157880/197816151-0c12cb15-9ad1-4932-84be-38e1cc1c8aaa.gif)


### Current behaviour


When `hasStickyHead` prop is true and rows are interactive and focused the header is overlapped when scrolling the table body.

No support for rendering right hand side sticky columns

### Checklist




#### QA


### Additional context


I have not added a cypress test as the work to refactor the tests for `FlatTable` is currently in flight, I will add one once that has been merged

### Testing instructions


Linked sandbox for issue, storybook stories. I will be happy to build sandboxes if QA request certain examples

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.
